File: 2021/09/11/main.go
package main

func calculate(s string) int {
	res := 0
	stack := make([]int, 0, len(s))
	sign := 1
	num := 0

	for i := 0; i < len(s); i++ {
		switch s[i] {
		case '1', '2', '3', '4', '5', '6', '7', '8', '9', '0':
			num = 0
			for ; i < len(s) && s[i] >= '0' && s[i] <= '9'; i++ {
				num = 10*num + int(s[i]-'0')
			}
			res += sign * num
			i--
		case '+':
			sign = 1
		case '-':
			sign = -1
		case '(':
			stack = append(stack, res, sign)
			res = 0
			sign = 1
		case ')':
			sign = stack[len(stack)-1]
			temp := stack[len(stack)-2]
			stack = stack[:len(stack)-2]
			res = sign*res + temp
		}
	}

	return res
}

File: 2021/09/11/main.py
class Solution:
    def calculate(self, s: str) -> int:
        num, res, sign, stack = 0, 0, 1, []

        for i in range(len(s)):
            c = s[i]
            if c in '0123456789':
                num = 0
                while i < len(s) and c >= '0' and c <= '9':
                    num = 10 * num + ord(c) - ord('0')
                    i += 1
                res += sign * num
                i -= 1
            if c == '+':
                sign = 1
            if c == '-':
                sign = -1
            if c == '(':
                stack.append(res, sign)
                res = 0
                sign = 1
            if c == ')':
                last = stack.pop()
                sign = last[1]
                res = sign * res + last[0]
                sign = 1

        return res